### PR TITLE
EEP 7: ETS Code Modernization

### PR DESCRIPTION
**Pull request by corranwebster**
A quick EEP on the proposed process of updating to remove Python 2 and start using features from Traits 6.1+

### CODE REVIEW

File: docs/source/eeps/eep-7.rst [hunk 1]
```diff
@@ -0,0 +1,110 @@
...
+:Content-Type: text/x-rst
+:Created: 2020-10-07
+:Post-History: 2020-10-07
+
+
+Introduction
+============
+
+All of the ETS codebase, with the exeption of the Codetools module,
+supports Python 3.  With the end-of-life of Python 2 in 2020, we want
```
Comment (rahulporuri) [inline docs/source/eeps/eep-7.rst:17]: codetools also supports python 3 - the last release is 4.3.0 which supports python 3.5 and 3.6 - https://github.com/enthought/codetools/blob/4.3.0/CHANGES.txt.

I don't think all parts of codetools support python 3.6 though.

File: docs/source/eeps/eep-7.rst [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+to move to a Python 3-only codebase.
+
+Additionally, we would like to make use of some of the new features of
+Traits 6.1 and remove things which were deprecated in Traits 6.0 and 6.1.
+In doing this we want to minimize the disruption to other codebases.
+
+This also presents a plan for dropping support for old versions of the
+GUI toolkits (Qt 4, WxPython < 4.1).
```
Comment (rahulporuri) [inline docs/source/eeps/eep-7.rst:25]: We can't drop support for Qt4 (PyQt4) yet because internal projects still rely on it.

File: docs/source/eeps/eep-7.rst [hunk 3]
```diff
@@ -0,0 +1,110 @@
...
+Motivation
+==========
+
+ETS is a codebase with roots that go back 15 years.  As a result there
+is a lot of old code that either doesn't take advantage of new features,
+or which is overly complicated because it needs to take into account the
+changes in behaviour of code over the past decade and a half.
+
+Currently the codebase supports Python 2.7 and 3.5+ in most cases, with
+the notable exception of Codetools.
```
Comment (rahulporuri) [inline docs/source/eeps/eep-7.rst:36]: see earlier comment about codetools supporting python 3.5/3.6

File: docs/source/eeps/eep-7.rst [hunk 4]
```diff
@@ -0,0 +1,110 @@
...
+- Apptools (current version 4.5.0)
+- Enable (current version 4.8.1)
+- Chaco (current version 4.8.0)
+- Mayavi (current version 4.7.2)
+- Scimath (current version 4.2.0)
+- Graphcanvas (current version 4.1.0)
+
+We might also use this process for Codetools, but the effort there is
+likely to be larger since it is more closely bound to the syntax of Python
+by the nature of the library.
```
Comment (rahulporuri) [inline docs/source/eeps/eep-7.rst:76]: PR https://github.com/enthought/codetools/pull/22 explicitly states what parts of codetools have been ported over and which haven't.

File: docs/source/eeps/eep-7.rst [hunk 5]
```diff
@@ -0,0 +1,110 @@
...
+All ETS projects which have not yet dropped Python 2 support will do so
+in their next major release, which is release 5.0 for all these libraries.
+
+We propose that the next major release of all ETS projects should work
+with Traits 6.0, 6.1 and 6.2.  This means that they will not be able to
+take advantage of new features of Traits, but hopefully smooths migration
+paths.
+
+The minor release after that, 5.1, should drop support for Traits 6.0 and
+6.1 and should as much as possible remove everything that has been
```
Comment (rahulporuri) [inline docs/source/eeps/eep-7.rst:92]: I'm not sure what you mean by dropping support for traits 6.1. Is the assumption here that traits 6.2 is already available?

File: docs/source/eeps/eep-7.rst [hunk 6]
```diff
@@ -0,0 +1,110 @@
...
+deprecated in Traits 6.0.
+
+We propose that for all ETS projects that have not yet dropped Python 2
+support that the next major release should support Qt 4.8, Qt 5.6+, and
+WxPython 4.0+ as much as possible.  This will likely require modernization
+of any WxPython code that these libraries contain, and possibly some minor
+fixes to Qt code, but is not likely to be a major effort.
```
Comment (rahulporuri) [inline docs/source/eeps/eep-7.rst:99]: It might be useful to elaborate a little on what exactly this work entails - which of the aforementioned packages have Qt4/Wx4 specific code, do they support both Qt4, Qt5 already?

And like I mentioned earlier, we can't drop support for Qt4 entirely because we still have projects relying on PyQt4.

File: docs/source/eeps/eep-7.rst [hunk 7]
```diff
@@ -0,0 +1,110 @@
...
+
+The releases for Pyface 7.2 and TraitsUI 7.2 should drop support for Qt
+versions earlier than 5.6 and WxPython 4.0.  This will likely have little
+downstream impact, but libraries might also choose to make this change in
+their 5.2 releases if they have Qt or Wx-specific code.
```
Comment (rahulporuri) [inline docs/source/eeps/eep-7.rst:104]: Additionally, Pyface 7.2 and TraitsUI 7.2 will be the last versions to support Traits 6.0 (and maybe even Traits 6.1).

File: docs/source/eeps/eep-7.rst [hunk 8]
```diff
@@ -0,0 +1,110 @@
...
+
+Finally, as part of code clean-up, re-formatting to match modern code
+style should be done opportunistically during the modernization process.
+In particular, bulk clean-up, like running ``black`` over a codebase should
+be done at a point where there are not a lot of open PRs, to reduce the
+impact from merge conflicts from re-formatting operations.
```
Comment (rahulporuri) [inline docs/source/eeps/eep-7.rst:110]: let's also explicitly specify that CI should be setup to ensure that the codebase stays flake8-clean.

File: docs/source/eeps/eep-7.rst (hunk 1 above)
Comment (corranwebster) [reply docs/source/eeps/eep-7.rst:17]: Yep, `codetools.blocks` and `execution` aren't ported and that's two-thirds of the functionality of the library.

File: docs/source/eeps/eep-7.rst (hunk 2 above)
Comment (corranwebster) [reply docs/source/eeps/eep-7.rst:25]: What's the blocker from them moving?  If we get them a 5.0 release with Qt4 support but they need to do some extra work to use new 5.1 features, that's OK I think.

If the answer needs to reference internal info which we shouldn't discuss here, ping me.

File: docs/source/eeps/eep-7.rst (hunk 4 above)
Comment (corranwebster) [reply docs/source/eeps/eep-7.rst:76]: Yes, `codetools.blocks` is the main blocker in that it needs a complete re-write for Python 3 and it is a huge part of the library.  It needs to be re-written to use (probably) the Python AST module.

File: docs/source/eeps/eep-4.rst [hunk 9]
```diff
@@ -56,9 +56,9 @@ to Traits, and would not permit existing code to enjoy the benefits of
 type annotation.
 
 Additionally, there has been substantial change in the design of standard
-library support for type annotations (particularly the :module:`typing`
+library support for type annotations (particularly the :mod:`typing`
```
Comment (mdickinson) [inline docs/source/eeps/eep-4.rst:59]: Could we please separate these changes out into another PR? They're unrelated to EEP 7, as I understand it.

File: docs/source/eeps/eep-7.rst (hunk 5 above)
Comment (corranwebster) [reply docs/source/eeps/eep-7.rst:92]: Yes, I think so; this is based on the stand-up yesterday.  I think that we are likely to get 5.0 releases out soon, but 5.1 releases are likely going to be after Traits 6.2 (I would expect the first 5.1 releases towards the end of the quarter, or early next year).

File: docs/source/eeps/eep-7.rst [hunk 10]
```diff
@@ -0,0 +1,110 @@
...
+:Type: Standards Track
+:Content-Type: text/x-rst
+:Created: 2020-10-07
+:Post-History: 2020-10-07
+
+
+Introduction
+============
+
+All of the ETS codebase, with the exeption of the Codetools module,
```
Comment (mdickinson) [inline docs/source/eeps/eep-7.rst:16]:
```suggestion
All of the ETS codebase, with the exception of the Codetools module,
```

File: docs/source/eeps/eep-7.rst (hunk 6 above)
Comment (corranwebster) [reply docs/source/eeps/eep-7.rst:99]: Probably the biggest effort is in Enable, as it has the most wx code.

Removing Qt4 support in most cases is a matter of just finding places where we are branching on the Qt version and removing the Qt4 branch.  It is a small effort.

File: docs/source/eeps/eep-7.rst [hunk 11]
```diff
@@ -0,0 +1,110 @@
...
+supports Python 3.  With the end-of-life of Python 2 in 2020, we want
+to move to a Python 3-only codebase.
+
+Additionally, we would like to make use of some of the new features of
+Traits 6.1 and remove things which were deprecated in Traits 6.0 and 6.1.
```
Comment (mdickinson) [inline docs/source/eeps/eep-7.rst:21]: May be worth reassuring people that none of the deprecated pieces will be removed from Traits until Traits 7.0.0 at the earliest (and in general, removal of deprecated items should only happen at a major version number bump).

File: docs/source/eeps/eep-7.rst (hunk 7 above)
Comment (corranwebster) [reply docs/source/eeps/eep-7.rst:104]: I didn't really want to worry about Traits/TraitsUI here as we've done most of the clean-up steps already.

File: docs/source/eeps/eep-7.rst (hunk 11 above)
Comment (corranwebster) [reply docs/source/eeps/eep-7.rst:21]: Yes, this is unclear.  I meant remove the _usage_ of deprecated things.

Comment (mdickinson) [reply docs/source/eeps/eep-7.rst:21]: Ah, sorry; I misunderstood.

File: docs/source/eeps/eep-4.rst (hunk 9 above)
Comment (corranwebster) [reply docs/source/eeps/eep-4.rst:59]: I've pulled these out and self-merged via a separate PR (they were in a seperate commit, so it was straightforward).

**Comment by corranwebster** [comment]
I think that everything is addressed here.

**Review by mdickinson — APPROVED** [review]
Looks good to merge to me.

To be clear to everyone about the EEP process: merging doesn't constitute acceptance. After it's merged, the next step in the process is to discuss whether the changes are a good idea or not.
